### ISSUE TITLE
Showing context variables never works in PHP7

### ISSUE THREAD

**Issue by zilti** [ahungry/geben#48]
It indefinitely stays stuck with the title line saying `(loading...)` at the end.

**Comment by ksjogo** [comment]
Afair we never made specific changes for PHP 7. Am currently not having any PHP projects, so no time to look into it.

**Comment by keisetsu** [comment]
I hope someone can fix, this makes geben and emacs unusable for php debugging.

**Comment by mallt** [comment]
I'm using php 7.2 with Xdebug 2.6.0 and the latest geben version from melpa (20170801.1251), and don't seem to have issues to display the context variables.

Do you see any special logging in the `*Messages*` buffer when trying to display the context variables?

**Comment by keisetsu** [comment]
Hmm, all the same versions but no luck, and no useful messages. I don't suppose php-mode version has anything to do with it? There seems to be a lot of work going on there. Just upgraded to           20191111.1650 to check, still not working.

**Comment by mallt** [comment]
I don't think the php-mode version should make a difference (I'm also using the latest one). Does `M-x toggle-debug-on-error` perhaps provide any useful information?

**Comment by keisetsu** [comment]
I've never used `toggle-debug-on-error` before, but trying it now doesn't do anything.

**Comment by keisetsu** [comment]
Also note that opening the variables screen kills geben. Even if I quit the variables view, geben just doesn't work anymore, I have to restart emacs. Or use Eclipse :(.

**Comment by ksjogo** [comment]
Is that with `emacs -Q` to disable local config?

**Comment by keisetsu** [comment]
How can I use geben with `emacs -Q`?

**Comment by mallt** [comment]
You can load geben from your package-user-dir in the scratch buffer f.ex:
```elisp
(load "/home/user/.emacs.d/elpa/geben-20170801.1251/dbgp")
(load "/home/user/.emacs.d/elpa/geben-20170801.1251/geben")
```

**Comment by keisetsu** [comment]
Yep, loaded in after `emacs -Q`, geben still fails to show variables. 

It looks like it is specific to a certain library, hoa/consistency. The variable viewer always fails and makes geben stall when it gets to this line https://github.com/hoaproject/Consistency/blob/fd7d0adc82410507f332516faf655b6ed22e4c2f/Prelude.php#L61.

The unicode periods of ellipsis looks odd, though the variable display is fine until after that is `$define`d

I don't know if this is just an example of a wider problem, or if it is something wrong with hoa/consistency. 

Note that geben works fine as long as I don't open the context view.

**Comment by mallt** [comment]
It seems the problem is related to the `…` character in the code you mention. The xdebug log shows a FAILED property name:
```
<- context_get -i 30035 -c 2 -d 0
-> <response xmlns="urn:debugger_protocol_v1" xmlns:xdebug="http://xdebug.org/dbgp/xdebug" command="context_get" transaction_id="30035" context="2"><property name="SUCCEED" fullname="SUCCEED" type="string" size="7" facet="constant" encoding="base64"><![CDATA[bGFsYWFhYQ==]]></property><property name="FAILED" fullname="FAILED" type="bool" facet="constant"><![CDATA[0]]></property><property name="…" fullname="…" type="string" size="6" facet="constant" encoding="base64"><![CDATA[bGFsYWxh]]></property></response>
```

I believe this is an encoding issue because xdebug seems to use iso-8859-1 in which the `…` character is not known, but I'm not sure if this can be fixed in geben or dbgp.

**Comment by keisetsu** [comment]
That's unfortunate. Thanks for your help.

**Comment by cimocimocimo** [comment]
Did anyone figure out a fix or hack to get this working? I'm having the same issue.

**Comment by ksjogo** [comment]
I suppose you could filter out variable lookups with an exclusion list.
Adding a  `defcustom` and one `unless` might be good enough.
Use `toggle-debug-on-error` to get the stack trace and the insertion point inside geben and go from there. 

Personally don't have a PHP setup anymore.

**Comment by cimocimocimo** [comment]
Aaaah, thanks for the tips. I'm not sure my emacs skills are up to the task so I'll just find other ways to get around this. I can use the eval statements if needed.

**Comment by keisetsu** [comment]
I need to get this working, so I'll give it a try and post any tips here if I get it to work.

**Comment by zilti** [comment]
Tbh at this point I am simply using `dap-mode` to debug PHP

**Comment by ahungry** [comment]
Why would anyone define a utf-8 constant is more puzzling to me (in that HOA lib).

Debugging this particular issue is a pain, I gave it a few hours of effort last night to no avail - even standard elisp mechanisms for setting debugger to pop up seem to offer no benefit, as the geben session just appears to stall out entirely on the line after it encounters a `define('<some unicode>` - I think the best/simplest would just be to have a middleware layer between the client and server dbgp that just strips out or safely encodes all utf8 chars to avoid current and future encoding issues globally.

**Comment by keisetsu** [comment]
For anyone specifically having a problem with hoa/consistency: I replaced the periods of ellipsis on line 60 with plain `...` and the pi character on line 72 with `pi`, and everything works again! It's not a perfect solution, but I'm very happy to have geben back - eclipse is so chunky and ugly!

**Comment by j-hotlink** [comment]
A fix for php 7.4 is available if you wish to port it: https://github.com/j-hotlink/geben

**Comment by ksjogo** [comment]
Seems reasonable to port these here.
May I ask why you forked the original version and not from here?

**Comment by j-hotlink** [comment]
I've been using the original codebase since 2010 and I'm not familiar with this repo - just sharing.

**Comment by brewers-dev-alex** [comment]
I'm so glad I found this thread. This issue has been bothering me for years. I applied the fix from j-hotlink's repo to my local copy of dgbp.el by simply swapping dbgp-session-response-to-chunk with the fixed version from this commit https://github.com/j-hotlink/geben/commit/ebef742f70e6cd559f1f5fd3965a78233942b249 and now I can display the context variables again.

**Comment by ahungry** [comment]
https://github.com/ahungry/geben/commit/dd05dc37d17947bb5bfbc1969ba6f0c8d5a43016

I applied the fix as stated, thanks @brewers-dev-alex and @j-hotlink for the fix!

If anyone can test, please let me know if this seems better for them as well, and we can close this old issue.

**Comment by ahungry** [comment]
This turned out to not work - dap-mode looks pretty awesome, if I have a need for interactive PHP/script debugging, I'll definitely give that a look.

I had to revert the j-hotlink fix, either it wasn't tested well, or I misapplied it.  In anycase, it left geben in a non-working state, which is now hopefully resolved via #56 - I'm going to close this, libraries shouldn't be using unicode IMO, and if we need to support it, we should have some type of filter on text in/text out that translates/encodes it vs messing with geben internals too much (unless at least 2+ people are willing to commit to the work+testing around it).
